base/rational.jl

```diff
@@ -20,13 +20,32 @@ struct Rational{T<:Integer} <: Real
         end
         return new(num2, den2)
     end
+
+    function Rational{T}(num::Integer, den::Integer, ::Val{false}) where T<:Integer
+        # Used when num and den are only known to be coprime and not both equal to 0
+        if T<:Signed && signbit(den)
+            den = -den
+            signbit(den) && __throw_rational_argerror_typemin(T)
+            num = -num
+        end
+        return new(num, den)
+    end
+
+    function Rational{T}(num::Integer, den::Integer, ::Val{true}) where T<:Integer
```

Comment:
How would you feel about making this a separate function called `unsafe_rational` instead? Just feels a tad more consistent to me here.

Reply:
I love the idea! I could definitely add an `unsafe_rational` function as a wrapper around this unsafe constructor, and exporting it would be consistent. However, I don't think it can replace the inner constructor here per se.
Indeed, there needs to be an inner constructor which bypasses all checks, and I can't change the original inner constructor otherwise that would be breaking. Since inner constructors have to have the name of the type (I think, is that right?), I have to add a constructor of the form `Rational{T}(...)`. Moreover, `Rational` is immutable, so I have to put the right numerator and denominator at construction time, so it has to take them as arguments. So the simplest idea to differentiate it from the original inner constructor simply consists in adding a `Val` parameter.

Does that make sense?

Reply:
> Since inner constructors have to have the name of the type (I think, is that right?), I have to add a constructor of the form Rational{T}(...).

It is actually possible to give the inner constructor a different name than the name of the struct, but you will probably have to use the workaround I outlined in #35498, because there are currently scoping issues.

Reply:
Interesting! I'm going to wait for comments on #35498 just to check how the core devs feel about that but otherwise that would be pretty neat.

Reply:
Implemented in the new commit. I really like it!